Sources/AuthFoundation/Token Management/Internal/KeychainTokenStorage.swift

```diff
@@ -87,69 +96,74 @@ class KeychainTokenStorage: TokenStorage {
-            .save()
+        let accessibility = security?.accessibility ?? .afterFirstUnlock
+        let accessGroup = security?.accessGroup
+        let accessControl = try security?.createAccessControl(accessibility: accessibility)
+        let newItem = Keychain.Item(account: id,
+                                    service: KeychainTokenStorage.serviceName,
+                                    accessibility: accessibility,
+                                    accessGroup: accessGroup,
+                                    accessControl: accessControl,
+                                    synchronizable: accessibility.isSynchronizable,
+                                    label: nil,
+                                    description: nil,
+                                    generic: nil,
+                                    value: data)
         
-        delegate?.token(storage: self, replaced: id, from: oldToken, to: token)
+        try? oldResult.delete()
+        try newItem.save(authenticationContext: security?.context)
+
+        delegate?.token(storage: self, replaced: id, with: token)
     }
     
     func remove(id: String) throws {
+        try Keychain
+            .Search(account: id,
+                    service: KeychainTokenStorage.metadataName)
+            .delete()
+        
         try Keychain
             .Search(account: id,
                     service: KeychainTokenStorage.serviceName)
-            .get()
             .delete()
 
-        let metadataItems = try Keychain
-            .Search(account: id,
-                    service: KeychainTokenStorage.metadataName,
-                    accessGroup: nil)
-            .list()
-        if !metadataItems.isEmpty {
-            try metadataItems.first?.get().delete()
-        }
-        
         delegate?.token(storage: self, removed: id)
 
         if defaultTokenID == id {
             try setDefaultTokenID(nil)
         }
     }
     
-    func get(token id: String) throws -> Token {
+    func get(token id: String, prompt: String? = nil, authenticationContext: TokenAuthenticationContext? = nil) throws -> Token {
         try token(with: try Keychain
                     .Search(account: id,
                             service: KeychainTokenStorage.serviceName)
-                    .get())
+                    .get(prompt: prompt,
+                         authenticationContext: authenticationContext as? KeychainAuthenticationContext))
     }
     
     func setMetadata(_ metadata: Token.Metadata) throws {
-        guard allIDs.contains(metadata.id) else {
-            throw TokenError.tokenNotFound(id: metadata.id)
-        }
-        
-        try Keychain
+        let item = Keychain
             .Item(account: metadata.id,
                   service: KeychainTokenStorage.metadataName,
                   accessibility: .afterFirstUnlock,
                   value: try encoder.encode(metadata))
-            .save()
+
+        try? item.delete()
```

Comment:
Risky. Maybe just build the logic around Add/Update APIs rather than Delete/Add? I think `Add/Update` is not that risky, at least can't cause data loss

Reply:
I mentioned this a bit up above, both the discussion around how to handle Delete/Add, as well as the use-cases around Update. For Metadata items, I'll change it to use Update, since I don't need to get the resulting item.